cloudify_marketplace/blueprint/scripts/configure.py

```diff
@@ -33,16 +52,251 @@ def regenerate_host_keys():
 
 
 def authorize_user_ssh_key(ssh_key):
+    ctx.logger.debug('Adding specified ssh key to authorized keys')
     with open('/home/<<IMAGEBUILDERUSER>>/.ssh/authorized_keys',
               'a') as auth_handle:
         auth_handle.write('{ssh_key}\n'.format(ssh_key=ssh_key))
 
 
+def get_auth_header(username, password):
+    header = None
+
+    if username and password:
+        credentials = '{0}:{1}'.format(username, password)
+        header = {
+            'Authorization':
+            'Basic' + ' ' + base64.urlsafe_b64encode(credentials)}
+
+    return header
+
+
+def build_certs(private_key_path,
+                public_key_path,
+                subjectaltnames,
+                openssl_conf_path='/etc/pki/tls/openssl.cnf'):
+    subjectaltnames = subjectaltnames.split(',')
+
+    common_name = subjectaltnames[0]
+    subjectaltnames = set(subjectaltnames)
+
+    subject_altdns = [
+        'DNS:{name}'.format(name=name)
+        for name in subjectaltnames
+    ]
+    subject_altips = []
+    for name in subjectaltnames:
+        ip_address = False
+        try:
+            socket.inet_pton(socket.AF_INET, name)
+            ip_address = True
+        except socket.error:
+            # Not IPv4
+            pass
+        try:
+            socket.inet_pton(socket.AF_INET6, name)
+            ip_address = True
+        except socket.error:
+            # Not IPv6
+            pass
+        if ip_address:
+            subject_altips.append('IP:{name}'.format(name=name))
+
+    subjectaltnames = ','.join([
+        ','.join(subject_altdns),
+        ','.join(subject_altips),
+    ])
+
+    subprocess.call([
+        'bash', '-c',
+        'openssl req -x509 -nodes -newkey rsa:2048 -keyout {private_key_path} '
+        '-out {public_key_path} -days 3650 -batch -subj "/CN={common_name}" '
+        '-reqexts SAN -extensions SAN -config <(cat {openssl_conf_path} '
+        '<(printf "[SAN]\nsubjectAltName={subjectaltnames}") )'.format(
+            private_key_path=private_key_path,
+            public_key_path=public_key_path,
+            common_name=common_name,
+            openssl_conf_path=openssl_conf_path,
+            subjectaltnames=subjectaltnames,
+        )
+    ])
+
+
+def get_host_ips():
+    # Get addresses for each interface that is active (up),
+    # and only for dynamic or permanent IPv6 addresses
+    ip_details = subprocess.check_output([
+        'ip', 'address', 'show', 'up', 'dynamic', 'permanent',
+    ])
+
+    # IP addresses are expected to be in one of these forms:
+    # inet <address>/<mask>
+    # inet6 <address>/<mask>
+    # We use a non-capturing group (?:) to ignore the inet/inet6 part
+    ip_finder = re.compile('(?:inet[6]? )([^/]+)')
+
+    return ip_finder.findall(ip_details)
+
+
+def get_general_subjectaltnames(subjectaltnames):
```

Comment:
Any ideas for a better name for this function or should we just drop the TODO and accept it as named?